frontend/src/framework/components/ParameterListFilter/private-utils/smartNodeSelectorUtils.ts

```diff
@@ -0,0 +1,179 @@
+    }
+    throw new Error(`Parameter type ${type} not supported`);
+}
+
+export function createAndAddNode(treeNodeDataList: TreeDataNode[], nodeName: string, icon?: string): TreeDataNode {
+    const newNode: TreeDataNode = { name: nodeName, description: "", icon: icon };
+    treeNodeDataList.push(newNode);
+    return newNode;
+}
+
+export function findOrCreateNode(treeNodeDataList: TreeDataNode[], nodeName: string, icon?: string): TreeDataNode {
+    const existingNode = treeNodeDataList.find((node) => node.name === nodeName);
+    if (existingNode) {
+        return existingNode;
+    }
+
+    return createAndAddNode(treeNodeDataList, nodeName, icon);
+}
+
+export function addParameterNameAndGroupToTreeDataNodeList(
+    treeNodeDataList: TreeDataNode[],
+    parameter: Parameter,
+    icon?: string
+): void {
+    // Parameter Name
+    const nameParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.NAME, icon);
+    if (!nameParentNode.children) {
+        nameParentNode.children = [];
+    }
+    findOrCreateNode(nameParentNode.children, parameter.name);
+
+    // Parameter Group
+    if (parameter.groupName) {
+        const groupParentNode = findOrCreateNode(treeNodeDataList, ParameterParentNodeNames.GROUP, icon);
+        if (!groupParentNode.children) {
+            groupParentNode.children = [];
+        }
+        findOrCreateNode(groupParentNode.children, parameter.groupName);
+    }
+}
+
+export function createTreeDataNodeListFromParameters(
+    parameters: Parameter[],
+    checkIcon?: string,
+    parentIcon?: string
+): TreeDataNode[] {
+    if (parameters.length === 0) {
+        return [];
+    }
+
+    const treeDataNodeList: TreeDataNode[] = [];
+
+    const hasContinuousParameter = parameters.some((parameter) => parameter.type === ParameterType.CONTINUOUS);
+
+    // Node for boolean/state properties on top level
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.CONTINUOUS, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.DISCRETE, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_CONSTANT, checkIcon);
+    createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_NONCONSTANT, checkIcon);
+    if (hasContinuousParameter) {
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LOGARITHMIC, checkIcon);
+        createAndAddNode(treeDataNodeList, ParameterParentNodeNames.IS_LINEAR, checkIcon);
+    }
+
+    // Add name and group for parameters
+    for (const parameter of parameters) {
+        addParameterNameAndGroupToTreeDataNodeList(treeDataNodeList, parameter, parentIcon);
+    }
+
+    return treeDataNodeList;
+}
+
+export function getParametersMatchingSelectedNodes(parameters: Parameter[], selectedNodes: string[]): Parameter[] {
+    // No selection implies no filtering
+    if (selectedNodes.length === 0) {
+        return parameters;
+    }
+
+    const delimiter = ":";
```

Comment:
Since the delimiter could be changed in the `SmartNodeSelector` component, maybe this should be an argument to the function with a default `:`?

Reply:
I considered this, but dropped it as this is private util, and the `SmartNodeSelector` is configured with the default delimiter.
I can add it for clarity, but if so I'll add a `const smartNodeSelectorDelimiter = ":"` in `ParameterListFilter` functional component s.t. the argument is actually used.

Reply:
I was only concerned about that somebody else might be editing your code and end up in a situation where things are not working again because they liked to change the delimiter. Makes our code more errorproof. I like your idea to put it into the functional component such that it is passed to both the `SmartNodeSelector` component and the utils function. Then it is very clear.